modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/GsubProcessor.java

```diff
@@ -49,7 +54,19 @@ String getReplacement() {
 
     @Override
     protected String process(String value) {
-        return pattern.matcher(value).replaceAll(replacement);
+        try {
+            return pattern.matcher(value).replaceAll(replacement);
+        } catch (StackOverflowError e) {
+            /*
+             * A bad regex on problematic data can trigger a StackOverflowError. In this case we can safely recover from the
+             * StackOverflowError, so we rethrow it as an Exception instead. This way the document fails this processor, but processing
+             * can carry on. The value would be useful to log here, but we do not do so for because we do not want to write potentially
+             * sensitive data to the logs.
+             */
+            String message = "Caught a StackOverflowError while processing gsub pattern: [" + pattern + "]";
+            logger.trace(message, e);
+            throw new ElasticsearchException(message);
```

Comment:
Can this be an IllegalArgumentException? I think if ElasticsearchException is used it will result in a 500 instead of a 4xx.

Reply:
It seems like a stretch of the definition of IllegalArgumentException, but I'm willing to do iot.

Reply:
It doesn't have to be IAE, but that maps to a 400. You could also create a new subclass of ElasticsearchException and specify the status explicitly, or look for other existing subclasses that return 4xx.